### PR TITLE
pypy: fix built-in library path

### PR DESCRIPTION
**Pull request by georgeharker**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

python has special patches for dyld to ensure homebrew libs get found.  This was absent for pypy / pypy3.  This PR fixes that.  In addition there was an issue with the ncurses ffi compilation which this PR also fixes.

### CODE REVIEW

**Comment by georgeharker** [comment]
I do not know why the SHA256 changed. Looking at downloads.python.org they list 837622130b36603a1893899bd9f529961a8e4a56c9eb67268d72ddf8920c9579 as the SHA256, which indeed it is.

**Comment by georgeharker** [comment]
rather than figure out the sha256 change I bumped to 7.3.16 as there's a new version anyway.  Verified that checksum against the one listed on the site, pypy2.7 was already on 7.3.16

File: Formula/p/pypy.rb [hunk 1]
```diff
@@ -81,6 +82,17 @@ def install
       s.gsub! "/include'", "/include/tcl-tk'"
     end
 
+    if OS.mac?
+      # Allow python modules to use ctypes.find_library to find homebrew's stuff
+      # even if homebrew is not a /usr/local/lib. Try this with:
+      # `brew install enchant && pip install pyenchant`
+      inreplace "lib_python/2.7/ctypes/macholib/dyld.py" do |f|
+        f.gsub! "DEFAULT_LIBRARY_FALLBACK = [",
+                "DEFAULT_LIBRARY_FALLBACK = [ '#{HOMEBREW_PREFIX}/lib', '#{Formula["openssl@3"].opt_lib}',"
```
Comment (carlocab) [inline Formula/p/pypy.rb:91]:
```suggestion
                "DEFAULT_LIBRARY_FALLBACK = [ '#{HOMEBREW_PREFIX}/lib',"
```
OpenSSL used to be keg-only, which is why we needed to add it there separately. But it's no longer keg-only so we don't need to do that anymore.

File: Formula/p/pypy3.10.rb [hunk 2]
```diff
@@ -72,6 +72,22 @@ def install
       s.gsub! "/include'", "/include/tcl-tk'"
     end
 
+    if OS.mac?
+      # Allow python modules to use ctypes.find_library to find homebrew's stuff
+      # even if homebrew is not a /usr/local/lib. Try this with:
+      # `brew install enchant && pip install pyenchant`
+      inreplace "lib-python/3/ctypes/macholib/dyld.py" do |f|
+        f.gsub! "DEFAULT_LIBRARY_FALLBACK = [",
+                "DEFAULT_LIBRARY_FALLBACK = [ '#{HOMEBREW_PREFIX}/lib', '#{Formula["openssl@3"].opt_lib}',"
```
Comment (carlocab) [inline Formula/p/pypy3.10.rb:81]:
```suggestion
                "DEFAULT_LIBRARY_FALLBACK = [ '#{HOMEBREW_PREFIX}/lib',"
```

File: Formula/p/pypy3.10.rb [hunk 3]
```diff
@@ -72,6 +72,22 @@ def install
...
+        f.gsub! "DEFAULT_FRAMEWORK_FALLBACK = [", "DEFAULT_FRAMEWORK_FALLBACK = [ '#{HOMEBREW_PREFIX}/Frameworks',"
+      end
+
+      # Fix mismatch in test definition of ncurses
+      inreplace "lib_pypy/_curses_build.py" do |f|
+        f.gsub! "char * tparm(const char *, ...);", "char * tparm(char *, ...);"
+      end
```
Comment (carlocab) [inline Formula/p/pypy3.10.rb:88]: Have you reported this upstream?

File: Formula/p/pypy3.10.rb (hunk 2 above)
Comment (georgeharker) [reply Formula/p/pypy3.10.rb]: thanks, removing

File: Formula/p/pypy3.10.rb (hunk 3 above)
Comment (georgeharker) [reply Formula/p/pypy3.10.rb:88]: I will do - I suspect it's specific to mac / the option which controls whether tparm is a varargs fn.  I need to figure out the right reporting mechanism for pypy

Comment (georgeharker) [reply Formula/p/pypy3.10.rb:88]: [resolved a](https://github.com/pypy/pypy/issues/4967)

**Comment by georgeharker** [comment]
Pypy says (re tparm issue):

This was done in 13e2004 last month, and will be part of the next release.

**Comment by georgeharker** [comment]
Any ideas here- thought I'd got everything. Works locally but all I can see are cancellations rather than actual errors

**Comment by georgeharker** [comment]
I'm going to try and split the two commits to avoid timeouts in tests

**Comment by georgeharker** [comment]
Second half is here https://github.com/Homebrew/homebrew-core/pull/172878

**Comment by georgeharker** [comment]
@Bo98 could you take a look at this related PR also?

**Review by carlocab — APPROVED** [review]
Thanks!

**Pull request merged**